### ISSUE TITLE
 Dove: We want to cheer... but have some mixed feelings

### ISSUE THREAD

**Issue by tvld** [feathersjs/feathers#2760]
## Our History

Our small team of eight has been using FeathersJs as Backend & API for about three years now. In the beginning, the first week we chose Feathers, was challenging. Feathers introduced new work and think patterns which took us some time to grasp. After that initial bump, we never looked back and started to recommend it to anybody. The community is very helpful and nice. It was great.

### Stability

The software proved super stable and all just works. We never had even one serious issue in all these years. I have tremendous admiration for the team behind the tool. Especially David, @daffl is by far the nicest, most knowledgable, and most responsive founder of an open source project that I know of. I am just very grateful that such people give their energy to the community.

### Nice to juniors

What is also great, is that junior developers could maintain our API. And this is a good thing when you realize that maintaining APIs is usually not the hobby of senior programmers. Of course, all juniors have this first tedious week where they need to get their heads around "hooks" and stuff before we can use Feathers, but it is worth it.

## Some minor wishes

During the time, we only had a few desires:

1. It would have been nice to have Typescript support. We moved to Vue 3 on the Front-End, and there we are happy to use TS
1. We would like to get rid of the ``myservice.class.js`` files from the ``feathers generate service``. You see, we had decided already about twenty(oops) years ago that software with abundant use of *classes* is not the easiest to maintain *over time*. It all sounds super savvy at first, but if the developers are either junior or even seniors when time passes, we better do without. I realize other people might feel different. But for us, when scaffolding new services, we just told the junior to pretend these files did not exist.
1. Documentation. We did not reach a level where we felt confident enough that we could contribute to the project with user manuals ourselves, but we knew we had to one day. Do not get me wrong, the current doc is great and we are thankful, but juniors almost always get a bit lost. They need to start Googling. Coincidentally, then there the best information again comes from David))

## Now Dove

So we were super happy to learn that in September 2022 we could start a new Greenfield project, and we decided to go for Feathers 5.0, aka Dove. However, truth be told, that has been a week of mixed feelings.

### The good

Typescript is omnipresent everywhere. You can not go wrong with that. Also, the project shows lots of love and care and feels robust and trustworthy.

### The bad

We found upgrading from previous Feathers to be less pleasant. Even though there is a  [long list of migration tips](https://dove.feathersjs.com/guides/migrating.html), we still get a lot of errors. For example, yesterday we noticed that no longer is working

```js
// in users.service.js
app.use('/version', { async find() { return { v: await versionInfo() } })
```
I know, these are small things. And while it will take us some weeks, I suppose it might make our code a bit even better.

### The Ugly

In all this, the main issue with Dove is that we feel we lost the beauty of the simplicity of Feathers.

As an example, there are no longer  ``myService.hooks.js`` files. This is sad, as it was a dream: simple files with the *before*, *after* and *error* hooks that we all understood. Perfect. Instead, Dove hides hooks in  ``myService.class.js``. And there it's confusing, as Dove introduces [around hooks](https://dove.feathersjs.com/guides/whats-new-in-dove.html#around-hooks). I read the documentation on that twice, but I can not get the logic. I am sure it is super savvy though.

Then the schemas. That's the worst about the new version. Previously a new `myService.model.js` was about ten lines of code, without field definitions. Now in the new `myService.schema.ts`, we have more than fifty lines! Of course, there all kinds of magic happen and no doubt the code is more robust for query and patch and whatnot. But so much boilerplate? For each new service again? It feels scary to me and to juniors at first glance.

The list is long, but that would be too much for this post.

## The future

I recall I felt the same long ago with Modx and later also with a full stack framework called Meteor. Both were amazing products created by, very smart developers. But after their initial success, new versions came that were *too* smart for the majority of the community. Both lost momentum and never really recovered.

*Maybe with Dove, it is just the CLI and scaffolding that tries to scare us. Maybe the elegant simplicity of the previous versions can be reinstated? What can we do to help?*

For now, it seems the best we can do is try to downgrade back to version 4.5, after the weekend though. It's Friday today))

About the long term, we do not know though. And hopefully, others can weigh in. What is wise? Should we accept that the target audience for Feathers has changed to a more sophisticated, higher level of developers? Should we stay on version 4? What about long-term maintenance?

**Comment by daffl** [comment]
Thank you for your feedback. This is always helpful. Here are my thoughts on your points, some of which are totally valid:

## Migration

If you are getting any errors migrating, please submit an issue. Anything that is not documented in the guide should still work as is especially all existing hooks and services. Your example should still work so it would be important for us to know if it doesn't.

## Hooks

The `around` hooks are based on https://github.com/feathersjs/hooks which we released a while ago. The basic thing that they allow is to give a single hook control of the entire before, after and error flow. As with everything in Feathers, they are entirely optional but used with the built-ins since some of those do need access to before and after at the same time. The reason why they are in the class file is that those kind of belong together and I wanted to reduce the number of files but maybe it makes more sense the other way around. We generally found that when using the new resolvers, there are a lot less hooks needed.

## Schemas

This is a complaint I have heard before and it'd probably make sense to investigate ways to reduce the setup a little bit more. The thing is that everybody will have to define their data model and validations somewhere somehow. So while everything in Feathers used to look super simple at the first glance, you end up getting that complexity somewhere else (how to convert the query, do validation, define and populate associations or get shared types were always big topics that didn't really have a good answer). Additionally, we found that - even though many of the things to watch out for are documented - especially Junior developers could end up building terribly insecure apps when everything "just works".

I look at schemas similar as to TypeScript which a large number of developers are now using because it makes things more safe even if it is quite a bit more verbose (coincidentally you get the correct types from schemas, too). In the long run it really reduced the code required for many things that were cumbersome in hooks. For example, this hook from the current chat guide to populate the user:

```js
module.exports = function (options = {}) { // eslint-disable-line no-unused-vars
  return async context => {
    // Get `app`, `method`, `params` and `result` from the hook context
    const { app, method, result, params } = context;
    // Function that adds the user to a single message object
    const addUser = async message => {
      // Get the user based on their id, pass the `params` along so
      // that we get a safe version of the user data
      const user = await app.service('users').get(message.userId, params);

      // Merge the message content to include the `user` object
      return {
        ...message,
        user
      };
    };

    // In a find method we need to process the entire page
    if (method === 'find') {
      // Map all data to include the `user` information
      context.result.data = await Promise.all(result.data.map(addUser));
    } else {
      // Otherwise just update the single result
      context.result = await addUser(result);
    }

    return context;
  };
};
```

Now looks like this as a resolver:

```ts
// Resolver for the data that is being returned
export const messagesResultResolver = resolve<MessagesResult, HookContext>({
  schema: messagesResultSchema,
  validate: false,
  properties: {
    user: async (_value, message, context) => {
      // Associate the user that sent the message
      return context.app.service('users').get(message.userId)
    }
  }
})
```

Maybe we can make generating the schema a separate step in the CLI. That way we basically have the old behaviour and then can introduce schemas. There are also still plans to allow customising the templates as well.

## Wishlist

1. I'm not sure I follow the TypeScript support comment.
2. This is doable. I was kind of hoping to put everything in a single file but then things got a little long for custom services. Unfortunately classes are the best tool to define services at the moment. I kind of blame TypeScript for this which works much better with classes and interfaces. In general, file structure things are always open for suggestions.
3. I am aware that the docs for this version still need a bunch of work but feedback like this is good to know what areas to focus on.

**Comment by jchamb** [comment]
Not trying to do the pile on, but also as a long time feathers users, I'd echo some of the above. I really like the foundational concepts being added and I don't really disagree with any of the approach. I just feel like the majority of what made it into Schema/Resolvers ended up adding a large amount of ambiguity for problems that using hooks could already solve, especially with the new `around` hooks. I have to agree that every dev on my team is gonna question what they should be using. From the docs when I first read this under "Result resolvers"

> or protect properties from being returned for external requests.

My first thought was like wait... what? We already have hooks for that stuff in hooks-common that are expressive and easy to understand at a glance. I was originally really excited for a more well defined resolvers to replace batchloaders and fastjoins, but where it landed in current form having `resolveResult `, `resolveDispatch ` `resolveAll` etc, is no where near as clear as before/after. Sure you can still get lost in hooks, but I would consider our app pretty large with well into the hundreds of services, shared and custom hooks all over. With some basic care and naming conventions it's not hard to grasp how the hooks work. I know for sure we'll be sticking with our old hooks to manage everything the resolver hooks are doing. That is totally okay, I don't expect feathers to be everything to everyone and solve all our problems. I am so ungodly thankful ya'll have created such a robust non-opinionated framework, where I can decide to not use it.

My only other complaint about schema is definitions being enforced through AJV/JSON Schema. This might just be me, I realize that it's popular, well defined, flexible and performant choice. In all honesty, if I was creating tooling for the masses I prob would have chosen it too. But... personally I still LOATH JSON schema definitions. It's verbose and clumsy for actual complex validations and to echo @tvld it has huge learning curve for our juniors. Our team chose Joi for its more naturally express definitions,  it's not for everyone and i'm not saying thats what should have been chosen.  I just really wish the schema/validate was provided more as general abstraction that you could plug and play whatever typing/validating library you wanted in. It seems unnatural for feathers to basically lock you to something when everywhere else there was so much care to not do that.

Everything else in Dove, i can't wait... well, except for Deno support, which isn't a grip, I just really don't care about it and hope that doesn't slow down final release.

I am curious, is sequelize being dropped from standard support, I see the docs only list knex now

**Comment by daffl** [comment]
I chatted with @marshallswain yesterday and I think we found a way to reduce the initial overhead/complexity of what is generated quite a bit more without loosing much of the flexibility. I'll also add the option to skip schema generation when creating a service (which everybody will have to use that at their own risk).

As for the resolvers - I'd look at it as a hook at the property level which just returns data and does not modify or create anything themselves. Where hooks are used for "side effects" like sending an email or logging etc. It helped us solve some pretty fundamental issues, like being able to safely populate associations and reliably define a safe data representation (without having to worry about passing around or checking for `params.provider`).

It would totally be possible to allow something like Joi instead of JSON schema for schema definition. In fact, that was what the very first version used. The problem was that frequent updates kept breaking the integration and then the maintainer stepped down leaving the project in a pretty uncertain state.

We pretty much evaluated every major library and I personally argued *against* JSON schema but was [convinced by many good points](https://github.com/feathersjs-ecosystem/schema/issues/20#issuecomment-678693674) being made in previous lengthy discussions. At least with a standard that is widely used you can rely on things not breaking and the tooling being much more solid than with anything library specific (at the time of my research, there e.g. wasn't a decent Joi to OpenAPI spec converter). What really turned me around was the great work @ThomasAribart has been doing with https://github.com/ThomasAribart/json-schema-to-ts - combined with resolvers, I don't believe anybody else has a similar approach for a database and ORM/library independent, type safe data processing layer (that can also be used on the client) - which is feels very much in the Feathers spirit.

Also, Deno support is tabled for now. It didn't seem to get a lot of interest let alone help with all the work that needs to be done - especially with the question out there if the Deno company might just come out with their own API framework (like they did with Fresh).

**Comment by marshallswain** [comment]
I'm adding my experience to the conversation. I get where people are coming from.  I started out pushing hard against pretty much everything in this release for the same reasons.  

### Things I was skeptical about

I was a major skeptic on `around` hooks back when they were still called `async` hooks.  I could see the value of giving a single hook access to the before and after parts of a request, but I LOVE our simple hooks API, and I typically get around the problem by storing something in `params` and retrieving it in the after hook.  Some endpoints end up storing a lot, but I was happy with the workflow because it worked and was familiar.

I didn't like Feathers Schemas because I loved my Joi tooling, as well. For all of the same reasons: it's easy to learn, easy to read, and easy to write.  Tough to beat.

I didn't like Feathers Resolvers because (1) I thought they were tied to the schemas (they're not), and (2) I was happy to use hooks like the hammer for all nails. ("When the only tool you have is a hammer, everything begins to look like a nail.") and (3) I looked at them as a completely separate and new API, not realizing the similarities (because I had no experience with them, yet).

## Things I learned

### About Feathers Resolvers

**The Feathers Resolvers are the most important part of this release.**  We just need to show simpler ways of using them.  The majority of the usage examples are based around generating a new app.  What's missing is the documentation showing how to implement them in an existing app.  I'll admit that this is mostly my fault.  The solution is pretty straightforward: I'll find places to document how to use them with existing APIs.  We need a resolvers guide.  The resolvers feel ambiguous, but once you start using them, there are very clear patterns of whether something works better in a resolver or in a hook.  A guide written for existing users will clear up most of the unknowns.

The dispatch resolvers are finally the solid solution to data leakage.  Using the `protect` hook, you had to be very careful when you populated data.  It has been too easy to forget that populating one more level would potentially expose sensitive data.  You had to be really careful to implement supportive patterns.  We needed a solution that was impossible to get wrong, and I feel like we've achieved that with the `resolveDispatch` hook.

One more note about resolvers. They kind of ARE a separate entity because we're going to end up with reusable resolver functions.  These functions will be far more reusable than hooks, with much less code.  The learning curve is actually VERY small.  The bigger frustration will come from wanting to upgrade our existing codebases and having to juggle a hook library with a resolver library.  I think I can help with this in the docs, as well.  For huge apps, you might want to just carry on with what you have already been doing.

### About Feathers Schemas

Feathers Schemas shine in three places for general users:

1. **A standards-compliant, officially-supported solution.**  It comes baked into the generator, now, so you don't have to go searching for solutions.  If we would have had this years ago, I wouldn't have created all of my tooling around Joi.  Back when JSON Schema was missing conditional schema support, it was too limited for what we needed.  Now it can support the huge majority of service APIs.  Having a standard around which we can build community tooling is something we've needed for a while.

2. **Offering a high degree of portability.**  It's one thing to publish a package with all of your Joi validations and install it in multiple places.  It's far more portable to be able to pass schemas around in network requests.  It's not a benefit that most apps will need, but having a standards-compliant schema that can be easily passed from worker to worker is pretty amazing for developing data-safe, dynamic apps.

3. **Existing as a machine-readable format.** JSON Schemas can be used in many tools for generating API docs.  I haven't done much experimentation in this area, because I tend to write a lot of documentation by hand.  But that's exactly why I like the idea.

If your app doesn't benefit from the above list, it's less likely to be worth the switch.  Feathers Schemas is 100% optional for most apps.

**Benefits for the people maintaining Feathers.**

We want to help people with their APIs, but there's still some fragmentation due to ORM lock-in.    Having an official solution replaces the scattered schemas/validations in most ORMs.  We can have conversations about a single validation solution, instead of having several sub-communities: one for each ORM.

**The rest of the benefits are generally "nice to haves".** 

They're more important for some than others.  I know I appreciate them, but they have less sway when it comes to tech decisions.

- Automated types from your schemas.  I love not having to create separate types.
- No more tight coupling with ORM validation.  The ORM solutions are pretty clunky compared to a standalone validation solution.  This benefit applies equally well to any standalone validation library.

### About `around` hooks

I wasn't a fan of `around` hooks until I converted most of my `before` and `after` hooks to resolver utilities.  When you're using regular hooks for everything, you need the granularity of `before` and `after` as opposed to one hook with access to both.  When you start using the resolvers, you don't need the granularity nearly as much.  Hooks become useful only for side effects outside of manipulating a record or populating data.  Things like caching and logging are great candidates to remain as hooks.

I admit that it took me over a year to like `around` hooks. @daffl and I had a lot of conversations about them and it took us a long time to figure out a way to integrate them in the most supportive way possible.

## Things I'm realizing

Now that I'm actually a fanboi of all of the new features, I see the problems falling into the following categories:

1. **We like the simplicity of what we're accustomed to doing.**

   It's fair to be concerned about new APIs, especially for somebody who is in charge of teaching those APIs to a team of developers.  I can see we need some more guide-centric material in the docs to help with this problem.  This is also part of the reason that I chose to make the docs for [Feathers-Pinia](https://v1.feathers-pinia.pages.dev) 100% guide-focused, with each guide containing a section for the relevant API docs.  We won't take that route with Feathers, but I now have some ideas for documentation improvements.

   **Action Item:** The solution to this issue can easily be to keep doing what you've been doing.  It's most likely that whatever you're already doing in your apps will continue to work, which is an often-overlooked, huge benefit that comes from this next point:

2. **There's not a lot of room for improvement in the existing APIs.  This is a testament to how good the Feathers core is.**

   The hooks API is so simple.  Once you get your head unstuck from the MVC patterns that dominate backend thinking, you fall in love with the cleanest implementation of aspect-oriented programming in API development (Accurate, at least, in any of the three languages I've used over the years: Ruby, PHP, Node).  Our hook object, with its declarative arrays of hooks, makes it awesome to work with AOP: instead of having a bunch of middleware scattered around the application, it's all in one place.
   
   I feel like @daffl and I found a good balance with the `around` hooks implementation.  They "play nice" with the existing hooks, which was priority number 1 for me.  So I say that there's not a lot of improvement in the existing APIs.  The hooks API is solid.  The addition of `around` hooks gives us an optional, cleaner way of handling before/after scenarios. But the resolvers do a great job of building ON TOP OF the existing API.

   **Action Item:** As with the previous item, the action step here is that we can keep using what we're already using.  There are optional tools available on top of the existing API.  Hooks are still the foundation.

3. **TypeScript can make things fugly.**  

   I don't think anybody in this current thread has this issue, but there are still a few others who do.  It's not an issue for those with TS experience, but for those with limited TS experience it can be daunting.

   With this release, we've gone all-in on TypeScript, and it adds a lot of extra code.  @daffl has expertly crafted the internal types.  He's skilled with types in a way that strikes a balance.  He made the internal Feathers types in a way that not many user-land types have to be added. A TypeScript app is always going to require more code.  Generate a JS app with the new generator and look at how terse and clean it is.  I miss those days, but the benefits of TypeScript are very apparent (as long as you don't let it slow down productivity, which takes practice).  

   Anyway, my opinion is that the TS code is definitely uglier, but better editor tooling with TS is a major win for junior devs.  And they can usually stick to using basic types when implementing hooks, etc.
   
   **Action Item:** There's not really an action item here.  We'll continue to do the best we can to keep types simple, focus on code, and only allow TS to improve our DX.

4. **The generated app is a bit too verbose**

   Building on the previous point, there's too much code in the current pre-release of the generated app.  With every new release of the generator, it takes some time to strike a balance between self-documenting and terse.  The current one leans too much towards advanced use cases.
   
   **Action item:** @daffl and I figured out some great ways to drastically simplify it.  It's much more Feathers-y, now  It's less intimidating, easier to understand, and offers a nice path to more advanced use cases.  We really appreciate the feedback that has led to the newest changes.
   
5. **We're failing to show the incremental upgrade path**

   This failing lands mostly on me for not noticing earlier.  I'm not really doing a good job of showing how to add resolvers to our existing apps.  The current docs and generator make it look like you need to use `around` hooks to use resolvers, but that's not accurate.

   Also, I'm not going a good job of relating to what we already know.  We are all accustomed to using utility hooks.  Resolvers are just another type of hook, but we talk about them like they're a completely standalone entity.  Resolvers are just a super-useful utility hook.  I can see that this is more of a communication problem than anything else.  Resolvers are powerful, and even junior-dev friendly.
   
   **Action Items:**
   - Docs for adding resolvers to an existing app.
   - A Resolver-specific migration guide to better illustrate what can be improved by a resolver.

**Comment by marshallswain** [comment]
@jchamb I personally loathe JSON Schema for probably the same reasons.  I don't think I'll ever get over it.  I'll write every line of code while gritting my teeth.  haha.

**Comment by marshallswain** [comment]
@tvld one thing you said which hasn't been addressed is about this snippet of code not working:

```ts
// in users.service.js
app.use('/version', { async find() { return { v: await versionInfo() } })
```

This code actually still works, but it demonstrates one of the ugly things about using TypeScript.  Your `ServiceTypes` object needs to be augmented.  Look at the bottom of `users.service.ts` and you'll see the declaration.  Instead of having to manually open a declarations file, the declaration is defined here and augments the one in the main `declarations` file. This is what it looks like in the generated code:

```ts
// Add this service to the service type index
declare module '../../declarations' {
  interface ServiceTypes {
    users: UsersService
  }
}
```

To support the line of code from your example, you need to augment the ServiceTypes.  The most JS-like way to solve the errors would be to declare the `/version` service as an `any` type:

```ts
// Add this service to the service type index
declare module '../../declarations' {
  interface ServiceTypes {
    users: UsersService
    '/version': any                  // here's the new `/version` service.
  }
}
```

You can also drop the `/` at the beginning and just do:

```ts
// in users.service.js
app.use('version', { async find() { return { v: await versionInfo() } })

// Add this service to the service type index
declare module '../../declarations' {
  interface ServiceTypes {
    users: UsersService
    'version': any                  // here's the new `/version` service.
  }
}
```

It'll work as long as the service path strings match.  Here's the full file:

![Screen Shot 2022-09-24 at 1 53 14 PM](https://user-images.githubusercontent.com/128857/192116078-4396c909-6ca5-43ad-9403-bea405722222.jpg)

**Comment by marshallswain** [comment]
Also @tvld, to directly answer your last few comments and questions:

> I recall I felt the same long ago with Modx and later also with a full stack framework called Meteor. Both were amazing products created by, very smart developers. But after their initial success, new versions came that were too smart for the majority of the community. Both lost momentum and never really recovered.

I experienced that pain from the Meteor community.  I was disappointed.  I've worked on Meteor apps in the last year and it really gave me an even greater appreciation for the simplicity of Feathers.

> Maybe with Dove, it is just the CLI and scaffolding that tries to scare us. Maybe the elegant simplicity of the previous versions can be reinstated? What can we do to help?

I believe this is the primary problem, yes.  The changes that @daffl and I discussed take your feedback into consideration.  We eliminated the class file.  We can't get rid of classes inside of the adapters.  They offer us some nice flexibility in creating adapters.  We agree that most apps don't need to customize classes.  We've reduced the class to a single line of code, which most devs will never need to touch.  For the ones who need it, it will be there.

The plan is to generate two files:
- the service file, which holds service setup and hooks
- one other file which will hold the schemas and resolvers, which have been greatly simplified.

So one file has everything that Feathers has always been.  The other file contains new, optional stuff.  If you want to keep chugging along the same as always, delete the second file and remove the hooks.

> For now, it seems the best we can do is try to downgrade back to version 4.5, after the weekend though. It's Friday today))

Yes.  I'd recommend trying again with the new generator, and please keep asking questions, especially if you can join us on Discord.

> About the long term, we do not know though. And hopefully, others can weigh in. What is wise? Should we accept that the target audience for Feathers has changed to a more sophisticated, higher level of developers? Should we stay on version 4? What about long-term maintenance?

The wisest path, from my perspective, would be to let us 
- make the changes based on the above feedback
- make updates to the docs for current apps to use new features.

Then give it another try.  There's no need to stay on version 4, since all of version 4 is still baked into 5.  We just need to do a better job of communicating with devs who have existing projects.  If your previous app is not TS-based, there will be some learning curve.  I'm doing my best to document TS-related issues here: https://dove.feathersjs.com/guides/troubleshooting.html.

As for feathers changing to a "higher level of developers", it might look like that on the surface because of the new tools.  The new APIs, in my opinion, will actually be much simpler and cleaner once you get accustomed to them.  Schemas might be an exception, depending on what you currently use.  I recommend trying out resolvers, first.  You're likely to appreciate them the most and wonder where they've been the last three years.  ;) 

Honestly, if you've been using only JS, previously, the jump to TS is the most annoying part.   You might try generating a JS-only app, first and trying out the new stuff after I write some more docs.

**Comment by marshallswain** [comment]
@jchamb I missed this line in the first read-through:

>  I know for sure we'll be sticking with our old hooks to manage everything the resolver hooks are doing.

With as large as your project is, I agree with that approach probably being the best way forward.  No need to refactor, especially with something that large.

**Comment by marshallswain** [comment]
@tvld, I've added your example registering services to the docs: https://dove.feathersjs.com/guides/troubleshooting.html#services-registering

**Comment by jchamb** [comment]
@marshallswain @daffl thanks for the responses. I love feathers not only because of how solid the framework is but also  because it's clear ya'll are deeply passionate about it's community. I see ya'll posting and answering basically everything in issue here and honestly don't know how ya'll manage to keep it feeling like you're just working on a project as a team with 2 or 3 other guys.  

As I said above I don't discredit any of the work ya'll have done and agree with the decisions. JSON Schema def wasn't the wrong call and I appreciate ya'll being able to have some long discussions setting aside personal tastes for what you believe will be the best for the community. I don't directly contribute, but I typically do read and follow the RFCs.

@marshallswain I'm not afraid to admit that I will definitely never get over it. I'll prob still be kicking and screaming over JSON schema definitions on my deathbed. 😛  

Starting a brand new project I don't think I would question any of the what Dove is bringing to the table. I mostly wanted to add some extra feedback and let @tvld know he's not alone in thinking how this all affects their team and that there is gonna be a lot of confusion for teams and developers that have had feathers as backbones of their apps for a while. 

@marshallswain you are most likely correct in diagnosing it as lack of clarity for transitioning some old patterns to new patterns. I do find it funny that you were against around hooks at first, my reaction was the opposite. Every time we have hooks signature with a before/after I'm always mumbly _...bleh fricken hack, meh its fine..._ `around` was an, Oh thank god, for me. They are the one thing I know we can use to simplify a lot of complexity around jamming things into params temporarily.

For the docs, here's some general thoughts from when I was reading through, which I hope will help in your improvements.

It might be wise to split resolvers away from schemas in the sectioning. All the current docs make it very much seem that they are tightly coupled together. Examples without schema usage, before introducing the thought that you can use schema here as well I think would help.

Better clarity on the definition of `resolvers`. It looks like the new definitions are more in line with what a GraphQL resolver would be.  My gut tells me though, that long time feathers users at first glance will confuse it with `resolvers` as part of `fastJoin` and think of it as just attaching related records/relationships outside of the `resolveData`. At the very least I'm pretty sure it'll confuse my team because we have a large reliance on fast joins and batch loaders. 

`resolveDispatch` I think has clarity for users that have a lot of DEEP experience in writing custom things for feathers, but I can tell you outside of myself there is maybe 1 other dev on my team that knows dispatch exists in the `context` in v4. I'd be curious to know how often across the community people are using dispatch directly to make that wording clear enough at a glance

It also isn't directly clear by just reading if the properties from property/result are always present or do you have to specifically ask for them like in GraphQL. Maybe I'm completely wrong about thinking of those resolve in a GraphQL like way and/or maybe I'm just to use to the `model` definitions being on at the DB layer.  

I've read the Resolver docs multiple times and it's not clear to me if Property or Result resolvers are the correct usage for relationships. It seems like I could use either. The Result resolvers I know specifically mentions populate associations but the property resolver example actually shows a data fetch. I haven't had the opportunity to actually read through the codebase yet to know directly, but it's unclear what kind of special care I need to take with data calls. I'm assuming resultResolvers are not doing the work of batching like BatchLoaders and you have to implement that? I think it'll be important to note what sort of handling needs done to avoid n+ queries based on result.

Side note if we still need to implement batch loaders, I'm still crossing my fingers ya'll have some new magic there. I don't even bother trying to explain implementing batch loaders to others. We've basically just spent the hours to build some advanced abstractions. We actually already have a `.resolvers.js` file for every service with the abstraction 😄 

Some more real world examples would be helpful as well. It's not immediately clear  how we could introduce resolvers dynamically without either bringing other hooks to the mix like `iff` or writing `if` logic based on context inside.

One thing we have in our codebase is a `$join` query param that allows us to dynamically include fastJoin resolvers when we actually want them. 

```
/users?$join[]=todos&$join[]=company
```

Thats where our `resolvers.js` comes in to play. Outside moving our split hook handling to an `around` hook. It's completely unclear how I could produce something similar with resolvers.


to be more specific with the schema definitions and locked in. What I specifically meant is for an example like this:
```
export const userSchema = schema({
  $id: 'UserData',
  type: 'object',
  additionalProperties: false,
  required: ['email'],
  properties: {
    email: { type: 'string' },
    password: { type: 'string' }
  }
} as const)

export const userDataResolver = resolve<User, HookContext>({
  schema: userSchema,
  validate: 'before',
  properties: {
    password: async (password) => hashPassword(password)
  }
})
```

I just wish there was some escape hatch for the resolver to use whatever schema you want like

```
export const userSchema = Joi.object({

    password: Joi.string()
        .pattern(new RegExp('^[a-zA-Z0-9]{3,30}$'))
        .required(),

    email: Joi.string()
        .email({ minDomainSegments: 2 })
        .required()
})

export const userDataResolver = resolve<User, HookContext>({
  schema: userSchema,
  validate: 'before',
  validateFn: (schema, data, context) => schema.validate(data),
  properties: {
    password: async (password) => hashPassword(password)
  }
})
```
If that does already exist in some shape or another making it clear in the docs would alleviate a lot of mystery and questioning around how someone would move a codebase to the new patterns where their schema/validation may not align with what's being provided as a default.

**Comment by marshallswain** [comment]
Thank you, @jchamb, for the considerate, detailed response.  I'll address more of the details later in the week, but wanted to point out that in version 5.1 the plan is to fully bring the new https://github.com/feathersjs-ecosystem/dataloader package into core.  We'll rewrite it in TS and test its limits.  I also want to build support for two populate query syntaxes, one similar to what I made with feathers-graph-populate and another following JSON API includes.

**Comment by tvld** [comment]
Dear friends, I am humbled by the extensive response. Although I feel I am a bit above my head. I will weigh a bit more...

### The Schema x Model Discussion

One of many beautiful, easy-to-understand mongoose models we have in a project:

```js
  const schema = new Schema({ 
    chatId: { type: String, maxlength: 80, ref: 'achievers', required: true },
    level: { type: Number, min: 0, max: 5, required: true, default: 0 },
    sphere: { type: String, enum: ['P', 'W'], maxlength: 1, required: true, default: 'P' },
    tags: [{ type: String, maxlength: 50, index: 1 }],
  }
```

It's so nice that this is what I always show juniors when introducing them to Feathers, in combination with before/after hooks that is. I do understand the choice for JSON Schema but it is a tough cookie to swallow, especially for juniors new to Typescript. 

So, I was thinking, can we offer a split-file approach? Where the model itself is simple as per above, and imported by the more complete schema file where all fancy things can happen?

> If splitting is not desired or would not make it more simple, maybe CLI offer "mongoose" database alternative to "mongodb" which would re-introduce the v4 schema / model format?

### Around hooks

For me, "around hooks" are still anti-pattern. You see, it seems to me that in v5 we have *two* times "before" and "after" code, instead of *one*:

```js
 // v4
 before[]
 after[]

 // v5
 before []
 after[]
 // and(!) / or:
 around [
  () => { 
    beforeCode
    await next()
    afterCode
  }
 ]

```

Although it is no reason to delay any Dove release I suppose, but maybe, if we could add the option to stick to *one* easy-to-understand paradigm, we can achieve the same:

```js
before: {
  all: [
    myAround
  ],
},
after: {
  all: [
    continueNext(myAround) // execute now the waiting one from a list
  ],
}

// ~/services/myService/hooks/myAround.js
module.exports = function () {
  return async function (context) {
    beforeCode
    await nextAfter()  // this pauses the function in a list of after hooks to continue as per above
    afterCode // executes only in after mode
  }
}

```

### Resolvers?

We LOVE "fastJoins", and I might misunderstand, but I have a hard time accepting if "resolvers" want to rob us of this beauty:

``` js
  joins: {
    accounts: () => async (user, context) => {
      user.accounts = (
        await context.app.service('accounts').find({
          query: { memberIds: user._id, $sort: {name: 1} }
        })
      ).data
    }
  }
```
I think documentation can indeed hand-hold a bit more, thanks @marshallswain ... ))

### Security

I understand what @daffl is saying "end up getting that complexity somewhere else". But the answer, IMHO, is *not* to force the security considerations *too early* in the development cycle.

You see, we often have no clue at first what to validate. In our daily practice, the limits imposed on input-data grow agile with the functionality of the app. Hence, I much rather see that we can start a simple mongoose or comparable simple ``model.js``. Then later we add validation and security considerations in separate files. Maybe with a default ``validate.js`` file?  This file we would create when the code is more or less working after our first app-iteration.

For Query parsing, I admit that we lose oversight at times with v4. Would it be possible to isolate query-check and builds an optional file as well?

@marshallswain writes "We needed a solution that was impossible to get wrong". I am not sure if that is needed. I think is ok to get it "wrong" first, if it gives speed and simplicity. Of course, it must be obvious that you need to make it right before deploying to production. Sometimes perfection is the enemy of progress...

### Typescript / "bug"

Wow, @marshallswain thank you for a very good explanation of my ``app.use('version')``. And yeah, it is my personal Typescript learning curve that caused this. but it is ok, the loss of clean code will be replaced by quality, and after the initial pain, juniors will appreciate it.

## Ideas to move on

### Documentation creation flow

There are so many beautiful text clips from senior Feathereers, here on GitHub, in Slack, and in Stackflow. What if they or the person asking (me;-) would make a Git issue with a special "*document suggestion*"-label and a link to that text? It would help to quickly assemble all stumble blocks.

### Functional files

More or less summarizing the ideas above: could CLI create this structure to keep all new Dove things but still make it accessible for newbies?

| myService file/folder | description | required
| --- |---| ---:|
| model.ts | *only* the Data model, in new JSON schema but visibly as clean as before | required
| schema.ts| the JSON schema for schema-fetishists, importing the ./model.ts, edit only in more advanced cases | required
| queryParser.ts | hardening, cleaning up or throwing faulty queries | optional
| validate.ts | more complex validation based on authorizations | optional
| resolvers.ts |  as per @jchamb, combining the slick "fastJoins" of v4  | optional
| hooks.ts | before/after/error | required
| .../hooks folder | opinionated structure to contain the external hook functions imported by hooks.ts | optional

### Final

After about four days, I feel that things are flowing nicely. Indeed as @marshallswain pointed out, the combination of a TypeScript learning curve and the about/resolvers/schema stuff, threw me off guard. Dove is starting to grow on me :-)))  When our first Dove project is finished, I will try to combine the text in this thread into a Medium post to share my experiences )))

**Comment by daffl** [comment]
Thank you everybody for the input! This discussion lead to some revisiting current v5 app patterns and solid refactoring and we tried to address many of the concerns mentioned here in https://github.com/feathersjs/feathers/pull/2772. More information can be found in the PR description but specifically working with @marshallswain and @fratzinger it includes the following:

- Reduced boilerplate for class and schema initialisation
- Hooks moved to the `<name>.ts` file where they are registered in a way that also makes them service type safe
- Resolver independent schema definitions
- Additional support for schema definitions using [TypeBox](https://github.com/sinclairzx81/typebox) which is more concise and declarative and similar to e.g. a Mongoose model definition
- Ability to generate a service without schemas

We are working on a detailed CLI guide that describes the patterns from the generator and how to do common things (like relationships, code formatting etc.).

**Comment by tvld** [comment]
Now, it has been a while, as private issues (Ukraine) took some attention. But, I am happy to say that today a complete re-do with .33 gave a big smile. Somehow the TypeBox makes a big difference. Thank you!  

And, I suppose it is ok to close this issue?...

A few first suggestions:

- I wonder if ``Omit`` instead of ``Pick`` for Schema's isn't a bit easier. Now we need to list all fields in a schema to allow ``create``:

```ts
// src/services/users/users.schema.ts
// ...
export const userDataSchema = Type.Omit(userSchema, ['_id'], { ...} // only need to list one field
// ...
```

- To keep hooks in an additional file is still preferred by us. The Hooks pattern is still one of the things that Feathers makes special IMHO. Most of the work we do is in the hooks file and not the service. So, it's no effort, but I think we keep isolating that code into the ``<service>.hooks.ts``, like:

```ts
// src/services/users/users.hooks.ts

import { authenticate } from '@feathersjs/authentication'
import { hooks as schemaHooks } from '@feathersjs/schema'
import { userDataValidator, userQueryValidator, userResolver, userDataResolver, userQueryResolver, userExternalResolver } from './users.schema'

export default {
  around: {
    all: [],
    find: [authenticate('jwt')],
    get: [authenticate('jwt')],
    create: [],
    update: [authenticate('jwt')],
    patch: [authenticate('jwt')],
    remove: [authenticate('jwt')]
  },
  before: {
    all: [
      schemaHooks.validateQuery(userQueryValidator),
      schemaHooks.validateData(userDataValidator),
      schemaHooks.resolveQuery(userQueryResolver),
      schemaHooks.resolveData(userDataResolver)
    ]
  },
  after: {
    all: [schemaHooks.resolveResult(userResolver), schemaHooks.resolveExternal(userExternalResolver)]
  },
  error: {
    all: []
  }
}


```
All in all: Big thumbs up for the great work :))))

**Comment by marshallswain** [comment]
Regarding Omit vs Pick vs alternatives: there are downsides to implementing each one.  I originally had the exact same thought as you, but as I've created more services I've noticed that Pick has worked for most of them.  With time we'll figure out what works best, though.
